File: src/leetcode/Javascript/724.Find-Pivot-Index.js
/**
 * Linear Search
 * Running Pointers Algorithms
 * Time O(N) | Space O(1)
 * https://leetcode.com/problems/find-pivot-index/
 * @param {number[]} nums
 * 
 * 
- traverse the elements - Linear loop(for loop) and get the total sum
- to calculate the right sum, compute below formula
(rightSum = totalSum - leftSum - current nums at index)
- this formala helps exclude the left sum and pivot index num to get the right sum 
- if condition allows us to compare the left and right sum 
 */
//Optimal Solution
const pivotIndex = (nums) => {
  const len = nums.length;
  let total = 0;
  let leftSum = 0;

  for (let i = 0; i < len; i++) {
    total += nums[i];
  }

  for (let i = 0; i < len; i++) {
    rightSum = total - leftSum - nums[i];
    if (rightSum === leftSum) {
      return i;
    }
    leftSum = leftSum + nums[i];
  }
  return -1;
};
module.exports = pivotIndex;

console.log(pivotIndex([1,7,3,6,5,6]))  //3
console.log(pivotIndex([1,2,3]))        //-1
console.log(pivotIndex([2,1,-1]))       //0

File: src/leetcode/python/724.Find-Pivot-Index.py
class Solution:  
    def pivotIndex(self, nums: List[int]) -> int:
        """
        Given an array of integers nums, calculate the pivot index of this array.
        The pivot index is the index where the sum of all the numbers strictly to the left of the index is equal to the sum of all the numbers strictly to the index's right.
        If the index is on the left edge of the array, then the left sum is 0 because there are no elements to the left. This also applies to the right edge of the array.
        Return the leftmost pivot index. If no such index exists, return -1.

        Solution:
        To solve this we have two running sums, an initial one that sums all elements to the end (right)
        which then decrements as the left sum increases.
        if the left and right sum is equivalent then return the current index
        """

        # initialize right sum
        right_sum = 0

        # populate right sum
        for x in nums:
            right_sum += x

        # initalize left sum
        left_sum = 0
        for i in range(len(nums)):

            # remove the value at possible pivot index from right sum
            right_sum -= nums[i]

            # compare for equality which would mean pivot index
            if left_sum == right_sum:
                return i
            # increate the left sum
            left_sum += nums[i]

        # if we get here we have no solution hence return -1
        return -1
